src/main/scala/is/hail/asm4s/FunctionBuilder.scala

```diff
@@ -85,110 +84,398 @@ abstract class FunctionBuilder[R](parameterTypeInfo: Array[TypeInfo[_]], returnT
+          }
+        }
+
+        f().asInstanceOf[R]
       }
     }
   }
+}
 
-  def newInstance[T]()(implicit tct: ClassTag[T], tti: TypeInfo[T]): Code[T] =
-    newInstance[T](Array[Class[_]](), Array[Code[_]]())
+trait ToI { def apply(): Int }
+class FunctionToIBuilder(implicit rti: TypeInfo[Int]) extends FunctionBuilder[Int](Array[TypeInfo[_]](), rti) {
 
-  def newInstance[T, A1](a1: Code[A1])(implicit a1ct: ClassTag[A1],
-    tct: ClassTag[T], tti: TypeInfo[T]): Code[T] =
-    newInstance[T](Array[Class[_]](a1ct.runtimeClass), Array[Code[_]](a1))
+  cn.interfaces.asInstanceOf[java.util.List[String]].add("is/hail/asm4s/ToI")
 
-  def newInstance[T, A1, A2](a1: Code[A1], a2: Code[A2])(implicit a1ct: ClassTag[A1], a2ct: ClassTag[A2],
-    tct: ClassTag[T], tti: TypeInfo[T]): Code[T] =
-    newInstance[T](Array[Class[_]](a1ct.runtimeClass, a2ct.runtimeClass), Array[Code[_]](a1, a2))
+  def result(c: Code[Int]): () => Int = {
+    val bytes = classAsBytes(c)
+    val localName = name.replaceAll("/",".")
 
-  def getArg[T](i: Int)(implicit tti: TypeInfo[T]): LocalRef[T] = {
-    assert(i >= 0)
-    assert(i < parameterTypeInfo.length)
-    new LocalRef[T](argIndex(i))
+    new Function0[Int] with java.io.Serializable {
+      @transient @volatile private var f: ToI = null
+      def apply(): Int = {
+        if (f == null) {
+          this.synchronized {
+            if (f == null) {
+              f = loadClass(localName, bytes).newInstance().asInstanceOf[ToI]
+            }
+          }
+        }
+
+        f()
+      }
+    }
   }
+}
 
-  def resultClass(c: Code[R]): Class[_] = {
-    mn.instructions.add(start)
-    c.emit(mn.instructions)
-    mn.instructions.add(new InsnNode(returnTypeInfo.returnOp))
-    mn.instructions.add(end)
+class Function1Builder[A >: Null, R >: Null](implicit act: ClassTag[A], ati: TypeInfo[A],
+  rti: TypeInfo[R]) extends FunctionBuilder[R](Array[TypeInfo[_]](ati), rti) {
 
-    val cw = new ClassWriter(ClassWriter.COMPUTE_MAXS +
-      ClassWriter.COMPUTE_FRAMES)
-    cn.accept(cw)
-    val b = cw.toByteArray
-    val clazz = loadClass(null, b)
-
-    // print bytecode for debugging
-    /*
-    val cr = new ClassReader(b)
-    val tcv = new TraceClassVisitor(null, new Textifier, new PrintWriter(System.out))
-    cr.accept(tcv, 0)
-    */
-
-    clazz
-  }
-
-  def whileLoop(condition: Code[Boolean], body: Code[_]): Code[Unit] = {
-    val l1 = new LabelNode
-    val l2 = new LabelNode
-    new Code[Unit] {
-      def emit(il: InsnList): Unit = {
-        il.add(l1)
-        condition.emit(il)
-        il.add(new LdcInsnNode(0))
-        il.add(new JumpInsnNode(IF_ICMPEQ, l2))
-        body.emit(il)
-        il.add(new JumpInsnNode(GOTO, l1))
-        il.add(l2)
+  cn.interfaces.asInstanceOf[java.util.List[String]].add("scala/Function1")
+
+  def arg1 = getArg[A](1)
+
+  def result(c: Code[R]): (A) => R = {
+    val bytes = classAsBytes(c)
+    val klass = act.runtimeClass
+    // FIXME: thread-safe?
```

Comment:
I think this is a hold over from before I got the double-checked locking thing all sorted out.